chain/src/txhashset/txhashset.rs

```diff
@@ -378,35 +376,63 @@ impl TxHashSet {
 			.backend
 			.check_compact(horizon_header.output_mmr_size, &rewind_rm_pos)?;
 
-		debug!("txhashset: compact height pos index...");
-		self.compact_height_pos_index(batch)?;
-
 		debug!("txhashset: ... compaction finished");
 
 		Ok(())
 	}
 
-	/// Initialize the output pos index based on current UTXO set.
-	/// This is a costly operation performed only when we receive a full new chain state.
+	/// (Re)build the output_pos index to be consistent with the current UTXO set.
+	/// Remove any "stale" index entries that do not correspond to outputs in the UTXO set.
+	/// Add any missing index entries based on UTXO set.
 	pub fn init_output_pos_index(
 		&self,
 		header_pmmr: &PMMRHandle<BlockHeader>,
-		batch: &mut Batch<'_>,
+		batch: &Batch<'_>,
 	) -> Result<(), Error> {
 		let now = Instant::now();
 
 		let output_pmmr =
 			ReadonlyPMMR::at(&self.output_pmmr_h.backend, self.output_pmmr_h.last_pos);
 
+		// Iterate over the current output_pos index, removing any entries that
+		// do not point to to the correct output.
+		let mut removed_count = 0;
+		for (key, (pos, _)) in batch.output_pos_iter()? {
+			if let Some(out) = output_pmmr.get_data(pos) {
+				if let Ok(pos_via_mmr) = batch.get_output_pos(&out.commitment()) {
+					if pos == pos_via_mmr {
+						continue;
+					}
```

Comment:
Is this cleanup foolproof? Lets imagine the following situation
- `output_pos` entry for commit `A` points to pos `x`
- `output_pos` entry for commit `B` points to pos `x`
- output MMR at pos `x` contains commit `B`

I don't think this loop will delete the incorrect entry for commit `A` from the `output_pos` db, because `pos == pos_via_mmr`.

Reply:
:+1: Yes I think you are right. We need to check both the pos matches _and_ the commitment matches here to `continue`.

Reply:
I had managed to convince myself that this was not necessary but I was not considering your scenario there.